azure-pipelines/clang-format-applied.yml

```diff
@@ -20,44 +20,16 @@ stages:
     steps:
     - bash: |
         wget -O - https://apt.llvm.org/llvm-snapshot.gpg.key | sudo apt-key add -
-        sudo apt-add-repository "deb http://apt.llvm.org/bionic/ llvm-toolchain-bionic-12 main"
+        sudo apt-add-repository "deb http://apt.llvm.org/bionic/ llvm-toolchain-bionic main"
         sudo apt-get update
-        sudo apt-get install -y clang-format-12
+        sudo apt-get install -y clang-format
       displayName: 'Install clang-format'
     - bash: |
         # Checkout the master branch as we require it to get the list of modified files
-        git checkout $(System.PullRequest.TargetBranch)
-        git fetch origin +$(Build.SourceBranch)
-
-        BASE_COMMIT=$(git merge-base FETCH_HEAD $(System.PullRequest.TargetBranch))
-        FILES=$(git --no-pager diff --name-only $BASE_COMMIT FETCH_HEAD -- '*.cpp' '*.hpp' '*.h')
-
-        echo "List of modified source files:"
-        for f in $FILES; do echo "$f"; done
-
-        # Checkout the PR so we can check the modified files for correct formatting
-        git checkout FETCH_HEAD
-
-        # Apply clang-format, modifying all badly formatted files
-        if [ ! -z "$FILES" ]; then
-            git --no-pager diff -U0 $BASE_COMMIT -- $FILES | perl azure-pipelines/util/format_diff_lines.pl
-        fi
+        git switch $(System.PullRequest.TargetBranch)
+        git switch $(System.PullRequest.SourceBranch)
+        BASE_COMMIT=$(git merge-base $(System.PullRequest.SourceBranch) $(System.PullRequest.TargetBranch))
+        echo $BASE_COMMIT
 
+        git clang-format --diff $BASE_COMMIT --
```

Comment:
`git clang-format` sometimes has a bug where exit code 1 is returned but there are no changes, e.g. stdout of this command is empty (I haven't found a reliable way to reproduce this bug, so I haven't reported it upstream yet). A check is needed for this case.

Source: I ran into this problem while setting up a clang-format check at work a few weeks ago...

Reply:
> We should update the VM image to Ubuntu 20.04, [which already has clang-format 12 installed](https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu2004-Readme.md#installed-software).

This method only works with llvm14+ https://github.com/llvm/llvm-project/commit/edbb8a843c130e60d71cb73e56a33d5ba2cc0ec9 https://github.com/llvm/llvm-project/issues/53220

But we could set the clang-format binary manually.

> Path include trigger should be set to *.c, *.cpp, *.h, *.hpp to avoid unnecessary runs of the format check.

Shouldn't be a problem: I think. git-clang-format and clang-format should already filter them, so that only compatible files are checked. But I can check this.

> Please attach a sample file (to be removed before merge) to test the format check.
Already done.

Reply:
There was a bug regarding the return code: https://github.com/llvm/llvm-project/issues/54758 Not sure if this was the reason.

Reply:
default extensions:   

```
default_extensions = ','.join([
      # From clang/lib/Frontend/FrontendOptions.cpp, all lower case
      'c', 'h',  # C
      'm',  # ObjC
      'mm',  # ObjC++
      'cc', 'cp', 'cpp', 'c++', 'cxx', 'hh', 'hpp', 'hxx',  # C++
      'ccm', 'cppm', 'cxxm', 'c++m',  # C++ Modules
      'cu', 'cuh',  # CUDA
      # Other languages that clang-format supports
      'proto', 'protodevel',  # Protocol Buffers
      'java',  # Java
      'js',  # JavaScript
      'ts',  # TypeScript
      'cs',  # C Sharp
      'json',  # Json
      ])
```
can be filtered with: 

```
The following git-config settings set the default of the corresponding option:
  clangFormat.binary
  clangFormat.commit
  clangFormat.extensions
  clangFormat.style
```

So this might be very straightforward.